### PR TITLE
Add error mapping

### LINKED ISSUE #1640: Add exception filter for our custom errors
**Which part of the project should be enhanced?**
Exception handling

**Is your enhancement request related to a problem? Please describe.**
Having long `catch` blocks everywhere makes files unnecessarily long and verbose and duplicates code.

**Describe the solution you'd like**
Implement a global [exception filter](https://docs.nestjs.com/exception-filters) that converts our custom errors like `AlreadyInDBError` to HTTP errors.

**Describe alternatives you've considered**
Continue having long `catch` blocks.

**Additional context**

### LINKED ISSUE #1983: We should not return Bad Request when an alias is already in use
### Which part of the project should be enhanced?

Both APIs

### Is your enhancement request related to a problem? Please describe.

Both the public and the private API return 400 Bad Request, when an alias is already used. This is a very generic error code.

### Describe the solution you'd like

The APIs should return 409 Conflict instead.

### Describe alternatives you've considered

Not changing the API.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by DerMolly**
### Component/Part
public and private api

### Description
This PR adds an error mapping from internal HedgeDoc Error to HttpExceptions. While doing this I noticed that some error codes are not documented in the openapi docs and added the appropriate annotations. 

### Steps

- [x] Added implementation
- [x] Added / updated documentation
- [x] I read the [contribution documentation](https://github.com/hedgedoc/hedgedoc/blob/develop/CONTRIBUTING.md) and
  made sure that:
  - My commits are signed-off to accept the DCO.
  - This PR targets the correct branch: `master` for 1.x & docs, `develop` for 2.x

Fixes #1983
Fixes #1640

### CODE REVIEW

File: test/private-api/alias.e2e-spec.ts [hunk 1]
```diff
@@ -83,7 +83,7 @@ describe('Alias', () => {
           .post(`/api/private/alias`)
           .set('Content-Type', 'application/json')
           .send(newAliasDto)
-          .expect(400)
+          .expect(409)
```
Comment (davidmehren) [inline test/private-api/alias.e2e-spec.ts:86]: Hmmm, actually, `Bad Request` seems to be a better error message for a forbidden note ID than `Conflict`. The best error would probably be `Forbidden` in this case.

File: test/private-api/alias.e2e-spec.ts [hunk 2]
```diff
@@ -156,7 +156,7 @@ describe('Alias', () => {
           .put(`/api/private/alias/${forbiddenNoteId}`)
           .set('Content-Type', 'application/json')
           .send(changeAliasDto)
-          .expect(400)
+          .expect(404)
```
Comment (davidmehren) [inline test/private-api/alias.e2e-spec.ts:159]: 404 when requesting a forbidden ID is also weird. IMO it should return `Forbidden`.

File: src/errors/error-mapping.ts [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+      'TokenNotValidError',
+      (object): HttpException => new UnauthorizedException(object),
+    ],
+    [
+      'TooManyTokensError',
+      (object): HttpException => new BadRequestException(object),
+    ],
+    [
+      'PermissionsUpdateInconsistentError',
+      (object): HttpException => new UnprocessableEntityException(object),
```
Comment (davidmehren) [inline src/errors/error-mapping.ts:51]: According to Wikipedia, HTTP 422 is WebDAV-specific

**Review by davidmehren — CHANGES_REQUESTED** [review]

File: src/errors/error-mapping.ts (hunk 3 above)
Comment (ErikMichelson) [reply src/errors/error-mapping.ts:51]: Wikipedia is right. HTTP 422 is defined in [RFC 4918](https://datatracker.ietf.org/doc/html/rfc4918#section-11.2) which is the RFC defining WebDAV.

File: test/private-api/alias.e2e-spec.ts (hunk 1 above)
Comment (ErikMichelson) [reply test/private-api/alias.e2e-spec.ts:86]: The question is: Do we want to let the end-user know which alias is forbidden?
In this case, I'd say okay 403 Forbidden is fitting.

But if we want no distinguishing between forbidden aliases and already taken aliases, it makes sense to map both to 409 Conflict.

File: test/private-api/alias.e2e-spec.ts (hunk 2 above)
Comment (ErikMichelson) [reply test/private-api/alias.e2e-spec.ts:159]: In this case 403 Forbidden seems right.

But speaking in strict terms it's "updating" and not "requesting" a forbidden alias.
For *requesting* a forbidden alias, 404 Not Found seems right.

File: src/api/private/alias/alias.controller.ts [hunk 4]
```diff
@@ -49,35 +56,32 @@ export class AliasController {
   ) {
     this.logger.setContext(AliasController.name);
   }
+
   @Post()
+  @ApiConflictResponse({ description: conflictDescription })
+  @ApiUnauthorizedResponse({ description: unauthorizedDescription })
+  @ApiNotFoundResponse({ description: notFoundDescription })
```
[the remark this answers is not in the document]
Comment (DerMolly) [reply src/api/private/alias/alias.controller.ts:63]: But `FullApi` is
```
ApiForbiddenResponse({ description: forbiddenDescription }),
ApiNotFoundResponse({ description: notFoundDescription }),
ApiUnauthorizedResponse({ description: forbiddenDescription }),
```

File: src/api/private/me/history/history.controller.ts [hunk 5]
```diff
@@ -26,11 +29,16 @@ import { SessionGuard } from '../../../../identity/session.guard';
...
+import {
+  notFoundDescription,
+  unauthorizedDescription,
+} from '../../../utils/descriptions';
 import { GetNoteInterceptor } from '../../../utils/get-note.interceptor';
 import { RequestNote } from '../../../utils/request-note.decorator';
 import { RequestUser } from '../../../utils/request-user.decorator';
 
 @UseGuards(SessionGuard)
+@UseFilters(ErrorExceptionMapping)
```
[the remark this answers is not in the document]
Comment (DerMolly) [reply src/api/private/me/history/history.controller.ts:41]: It wasn't always. This is remian of that. Fixed

File: src/errors/error-mapping.ts (hunk 3 above)
Comment (DerMolly) [reply src/errors/error-mapping.ts:51]: Well that's not good then, should we replace it with `BadRequest` or something else?

File: src/api/private/alias/alias.controller.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (davidmehren) [reply src/api/private/alias/alias.controller.ts:63]: Right. What I'd like is to merge each regularly used combination of individual decorators into a new one to reduce clutter.

File: src/errors/error-mapping.ts (hunk 3 above)
Comment (DerMolly) [reply src/errors/error-mapping.ts:51]: `BadRequest` will be used

File: src/api/private/me/history/history.controller.ts [hunk 6]
```diff
@@ -42,84 +50,57 @@ export class HistoryController {
   }
 
   @Get()
+  @ApiUnauthorizedResponse({ description: unauthorizedDescription })
+  @ApiNotFoundResponse({ description: notFoundDescription })
```
[the remark this answers is not in the document]
Comment (DerMolly) [reply src/api/private/me/history/history.controller.ts:51]: This will be fixed in a further PR

**Review by davidmehren — APPROVED** [review]

File: src/api/private/auth/auth.controller.ts [hunk 7]
```diff
@@ -49,55 +52,38 @@ export class AuthController {
 
   @UseGuards(RegistrationEnabledGuard)
```
Comment (ErikMichelson) [inline src/api/private/auth/auth.controller.ts:53]: This guard can throw a BadRequestException which should get an ApiBadRequestResponse decorator.

**Comment by DerMolly** [comment]
Annotations will be improved by #2076

File: src/api/private/auth/auth.controller.ts (hunk 7 above)
Comment (DerMolly) [reply src/api/private/auth/auth.controller.ts:53]: fixed

**Review by ErikMichelson — APPROVED** [review]

**Review by InnayTool — APPROVED** [review]

**Pull request merged**